streams.cpp

```diff
@@ -258,3 +300,248 @@ DS::Blob DS::HexDecode(const ST::string& value)
+}
+
+std::optional<DS::EncryptedStream::Type> DS::EncryptedStream::CheckEncryption(const char* filename)
+{
+    DS::FileStream stream;
+    stream.open(filename, "r");
+    return CheckEncryption(&stream);
+}
+
+std::optional<DS::EncryptedStream::Type> DS::EncryptedStream::CheckEncryption(DS::Stream* stream)
+{
+    uint32_t pos = stream->tell();
+    if (pos != 0)
+        stream->seek(0, SEEK_SET);
+    uint8_t header[12];
+    stream->readBytes(header, sizeof(header));
+    stream->seek(pos, SEEK_SET);
+
+    if (memcmp(header, "whatdoyousee", sizeof(header)) == 0)
+        return DS::EncryptedStream::Type::e_xtea;
+    if (memcmp(header, "BriceIsSmart", sizeof(header)) == 0)
+        return DS::EncryptedStream::Type::e_xtea;
+    if (memcmp(header, "notthedroids", sizeof(header)) == 0)
+        return DS::EncryptedStream::Type::e_btea;
+    return std::nullopt;
+}
+
+void DS::EncryptedStream::setKeys(const uint32_t* keys)
+{
+    memcpy(m_key, keys, sizeof(m_key));
+}
+
+void DS::EncryptedStream::bteaDecipher(uint32_t* buf, uint32_t num) const
+{
+    uint32_t key = ((52 / num) + 6) * 0x9E3779B9;
+    while (key != 0) {
+        uint32_t xorkey = (key >> 2) & 3;
+        uint32_t numloop = num - 1;
+        while (numloop != 0) {
+            buf[numloop] -=
+              (((buf[numloop - 1] << 4) ^ (buf[numloop - 1] >> 3)) +
+              ((buf[numloop - 1] >> 5) ^ (buf[numloop - 1] << 2))) ^
+              ((m_key[(numloop & 3) ^ xorkey] ^ buf[numloop - 1]) +
+              (key ^ buf[numloop - 1]));
+            numloop--;
+        }
+        buf[0] -=
+          (((buf[num - 1] << 4) ^ (buf[num - 1] >> 3)) +
+          ((buf[num - 1] >> 5) ^ (buf[num - 1] << 2))) ^
+          ((m_key[(numloop & 3) ^ xorkey] ^ buf[num - 1]) +
+          (key ^ buf[num - 1]));
+        key += 0x61C88647;
+    }
+}
+
+void DS::EncryptedStream::bteaEncipher(uint32_t* buf, uint32_t num) const
+{
+    uint32_t key = 0;
+    uint32_t count = (52 / num) + 6;
+    while (count != 0) {
+        key -= 0x61C88647;
+        uint32_t xorkey = (key >> 2) & 3;
+        uint32_t numloop = 0;
+        while (numloop != num - 1) {
+            buf[numloop] +=
+              (((buf[numloop + 1] << 4) ^ (buf[numloop + 1] >> 3)) +
+              ((buf[numloop + 1] >> 5) ^ (buf[numloop + 1] << 2))) ^
+              ((m_key[(numloop & 3) ^ xorkey] ^ buf[numloop + 1]) +
+              (key ^ buf[numloop + 1]));
+            numloop++;
+        }
+        buf[num - 1] +=
+          (((buf[0] << 4) ^ (buf[0] >> 3)) +
+          ((buf[0] >> 5) ^ (buf[0] << 2))) ^
+          ((m_key[(numloop & 3) ^ xorkey] ^ buf[0]) +
+          (key ^ buf[0]));
+        count--;
+    }
+}
+
+void DS::EncryptedStream::xteaDecipher(uint32_t* buf) const
+{
+    uint32_t second = buf[1], first = buf[0], key = 0xC6EF3720;
+
+    for (size_t i = 0; i < 32; i++) {
+        second -= (((first >> 5) ^ (first << 4)) + first)
+                ^ (m_key[(key >> 11) & 3] + key);
+        key += 0x61C88647;
```

Comment:
Could you flip the sign here so the magic number matches the client?

```suggestion
        key -= 0x9E3779B9;
```